### PR TITLE
DSG-886/Add-new-fields-to-user-profile

### PR DESCRIPTION
**Pull request by afcalderonm**
## Description
drop DisplayName and add firstName, middleName, lastName and phoneNumber to the userEntity

## Motivation and Context
To better support user profile management

## How Has This Been Tested?
tested modifying the automated test that already exist and also tested manually from postman

## Types of changes

Patch
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation change (documentation only, will not require redeployment)
- [x] Refactor or cleanup (functionality unchanged)
Minor
- [ ] New feature (non-breaking change which adds functionality)
Major
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

- [x] I have confirmed all acceptance criteria defined by the story have been met

### CODE REVIEW

File: acceptance-tests/src/functionalTest/resources/io/nuvalence/platform/features/RoleManagement.feature [hunk 1]
```diff
@@ -59,8 +59,8 @@ Feature: Role Management
       | ta2:basic   |
       | oa2:basic  |
     And existing application users with the following attributes
-      | externalId | displayName | email          |
-      | test-user  | Test User   | user@test.test |
+      | externalId | firstName | email          |
+      | test-user  | Test   | user@test.test |
     When user with externalId "test-user" is assigned role "test-role-2"
     And a GET request is made for application roles for the user with externalId "test-user"
```
Comment (asultan001) [inline acceptance-tests/src/functionalTest/resources/io/nuvalence/platform/features/RoleManagement.feature:65]: We probably want to add similar tests for other other properties or enhance this test to include the other properties?

File: swagger.yaml [hunk 2]
```diff
@@ -756,14 +761,22 @@ components:
       type: object
       required:
         - email
-        - displayName
         - externalId
         - identityProvider
         - userType
       properties:
-        displayName:
```
Comment (deibys-nuvalence) [inline swagger.yaml:764]: There are serveral other services (exampple work manager ) which query user-management to get displayName , since it is gone , in order to keep backwards compatibility, you should:
1. Change user-management client in other services
2. At least in the api response , keep the field and make it the union of firstname + middleName + lastname
3. FE will query user management to get the names to display per user

File: swagger.yaml [hunk 3]
```diff
@@ -756,14 +761,22 @@ components:
...
+          description: User's first name.
+        middleName:
+          type: string
+          description: User's middle name.
+        lastName:
+          type: string
+          description: User's last name.
+        phoneNumber:
+          type: string
+          description: User's phoneNumber.
```
Comment (asultan001) [inline swagger.yaml:779]: separate the word phone number in the description

**Comment by asultan001** [comment]
@afcalderonm can we add a few tests that validate that you can use null values here as well? I believe these properties should be optional.

File: service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java [hunk 4]
```diff
@@ -64,9 +64,10 @@ public UserEntity updateUserById(UUID userId, UserUpdateRequest updateRequest) {
                         .findByIdLoaded(userId)
                         .orElseThrow(() -> new ResourceNotFoundException("User not found."));
 
-        // Update displayName, externalId, and email if provided
-        updateDisplayName(userEntity, updateRequest);
+        // Update email if provided
+        updateNames(userEntity, updateRequest);
         updateEmail(userEntity, updateRequest);
+        userEntity.setPhoneNumber(updateRequest.getPhoneNumber());
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java:70]: Maybe it is not in requirements, but would be nice to validate a standard formatting for phone numbers
A library that can be helpful:
https://mvnrepository.com/artifact/com.googlecode.libphonenumber/libphonenumber

File: service/src/test/java/io/nuvalence/user/management/api/service/auth/UsersApiDelegateImpl.java [hunk 5]
```diff
@@ -79,7 +79,10 @@ private UserCreationRequest createNewUserModel() {
         UserCreationRequest user = new UserCreationRequest();
         user.setIdentityProvider("https://securetoken.google.com/my-project");
         user.setExternalId("SMGjTO5n3sZFVIi5IzpW2pI8vjf1");
-        user.setDisplayName("Rusty Popins");
```
Comment (deibys-nuvalence) [inline service/src/test/java/io/nuvalence/user/management/api/service/auth/UsersApiDelegateImpl.java:82]: Renaming this class to follow standard would be nice
I mean appending "Test" to class name

File: service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java [hunk 6]
```diff
@@ -64,9 +64,10 @@ public UserEntity updateUserById(UUID userId, UserUpdateRequest updateRequest) {
                         .findByIdLoaded(userId)
                         .orElseThrow(() -> new ResourceNotFoundException("User not found."));
 
-        // Update displayName, externalId, and email if provided
-        updateDisplayName(userEntity, updateRequest);
+        // Update email if provided
+        updateNames(userEntity, updateRequest);
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java:68]: Not sure if it is a requirement, but would be good to validate at least one of the names fields is set

File: swagger.yaml (hunk 2 above)
Comment (asultan001) [reply swagger.yaml:764]: It's not a bad idea to maintain the display name as a combination of the other names for backwards compatibility

File: service/src/main/java/io/nuvalence/user/management/api/service/cerbos/CerbosClient.java [hunk 7]
```diff
@@ -268,7 +268,10 @@ private CheckResourceSetRequest buildCheckRequest(
                 .resource(
                         ResourceSet.builder()
                                 .kind(resourceName)
-                                .instances(Map.of(userEntity.getDisplayName(), new AttributesMap()))
+                                .instances(
```
Comment (JayStGelais) [inline service/src/main/java/io/nuvalence/user/management/api/service/cerbos/CerbosClient.java:271]: This class, `CerbosClient` should have been removed when we removed the Controller methods that called it. This has been replaced by the generic access control services brought in with the `dsgov-auth` library.

File: service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java (hunk 6 above)
Comment (JayStGelais) [reply service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java:68]: Freshly minted users may not have any of the 3 name fields set as we can't depend on Identity Providers to have a value.

For example, Google Identity Platform email/password user pools only have an email address to describe the user.

File: service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java (hunk 4 above)
Comment (JayStGelais) [reply service/src/main/java/io/nuvalence/user/management/api/service/service/UserService.java:70]: Lets pause on validation of phone numbers for now. 

In general we handle formatting on the front end, and just store a string of digits internally. Any validation on what lenght that string needs to be would need to take into account the country involved. I thik it would be better for this solution to be flexible with what it can accept and address validations to meet specific client requirements as they come up in engagements.

File: service/src/test/java/io/nuvalence/user/management/api/service/service/CerbosClientTest.java [hunk 8]
```diff
@@ -163,7 +163,7 @@ private UserEntity getUserEntity() {
         role.setName(allowRole);
 
         var userEntity = new UserEntity();
-        userEntity.setDisplayName(allowRole);
+        userEntity.setFirstName(allowRole);
```
Comment (JayStGelais) [inline service/src/test/java/io/nuvalence/user/management/api/service/service/CerbosClientTest.java:166]: We should also delete this test, in addition to the class it is testing.

File: swagger.yaml (hunk 2 above)
Comment (JayStGelais) [reply swagger.yaml:764]: For now including displayName on the API model and have it concatenate `firstName + " " + lastName`.

There are some cultures where this logic is incorrect, but it is a worst a temporary fix for backwards compatibility.

File: acceptance-tests/src/functionalTest/resources/io/nuvalence/platform/features/RoleManagement.feature (hunk 1 above)
Comment (afcalderonm) [reply acceptance-tests/src/functionalTest/resources/io/nuvalence/platform/features/RoleManagement.feature:65]: i just modify the test to include the other properties

File: swagger.yaml (hunk 3 above)
Comment (afcalderonm) [reply swagger.yaml]: fixed

File: service/src/test/java/io/nuvalence/user/management/api/service/auth/UsersApiDelegateImpl.java (hunk 5 above)
Comment (afcalderonm) [reply service/src/test/java/io/nuvalence/user/management/api/service/auth/UsersApiDelegateImpl.java:82]: fixed

File: service/src/main/java/io/nuvalence/user/management/api/service/cerbos/CerbosClient.java (hunk 7 above)
Comment (afcalderonm) [reply service/src/main/java/io/nuvalence/user/management/api/service/cerbos/CerbosClient.java]: the class was removed

File: service/src/test/java/io/nuvalence/user/management/api/service/service/CerbosClientTest.java (hunk 8 above)
Comment (afcalderonm) [reply service/src/test/java/io/nuvalence/user/management/api/service/service/CerbosClientTest.java]: the test class was deleted

File: swagger.yaml (hunk 2 above)
Comment (afcalderonm) [reply swagger.yaml:764]: the displayName was included concatenated with firstName + " " + lastName

**Comment by afcalderonm** [comment]
> @afcalderonm can we add a few tests that validate that you can use null values here as well? I believe these properties should be optional.

i just add a few test to validate its possible to update the user profile with the new values null and also to get a user that has those values null

**Review by JayStGelais — CHANGES_REQUESTED** [review]

File: service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java [hunk 9]
```diff
@@ -218,6 +219,11 @@ private UserDTO mapUserEntity(UserEntity user) {
         userDto.setPreferences(
                 UserPreferenceEntityMapper.INSTANCE.convertUserPreferenceEntityToUserModel(
                         user.getUserPreference()));
+        userDto.setDisplayName(
```
[the remark this answers is not in the document]
Comment (asultan001) [reply service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java:222]: I think we should trim this operation as well so that we can avoid the following:

Firstname but null lastime -> "Abe " 
Lastname but null firstname -> " Sultan"
null firstname and lastname -> " " 

Instead if you trim them it should be "Abe", "Sultan", ""

File: service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java [hunk 10]
```diff
@@ -218,6 +218,8 @@ private UserDTO mapUserEntity(UserEntity user) {
         userDto.setPreferences(
                 UserPreferenceEntityMapper.INSTANCE.convertUserPreferenceEntityToUserModel(
                         user.getUserPreference()));
+        userDto.setDisplayName(UserEntityMapper.INSTANCE.createDisplayname(user));
```
Comment (JayStGelais) [inline service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java:221]: What I meant by the previous comment was that this should happen automatically when mapping a UserEntity to a UserDTO. This could be done by adding the followin annotation to `convertUserEntityToUserModel`

`@Mapping(target = "displatName", expression = "java(createDisplayname(userEntity))")`

**Review by JayStGelais — CHANGES_REQUESTED** [review]

File: service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java (hunk 10 above)
Comment (afcalderonm) [reply service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java]: oh, i did not know you can do that, now is applied

**Review by JayStGelais — APPROVED** [review]
